Adding xdaq control

jotadram6 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Adding new feature to create SID only if the run type is local

A new parameter StringT has been added: RUN_TYPE
If it is local, a SID is created, if not SID=0

For the moment this parameter is hardcoded and local by default. Will need to find a better way to configure it on the go.

Accordingly the session ID is only closed if running in local mode.
<!--- Describe your changes in detail -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue(s) here. -->
Changes needed to run in global mode.
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
It has been tested at 904 with uFEDKIT + TCDS config.

It has been tested that when moving to global config no new session is created.

### Screenshots (if appropriate):

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- Template thanks to https://www.talater.com/open-source-templates/#/page/99 -->

jsturdy commented on gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java:
```diff
@@ -4,10 +4,19 @@
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.util.Random;
+import java.net.URL;
```
There's a double import here, probably best to order these either by purpose or alphabetically to simplify this in the future (should also make sure we're not over-importing)

jsturdy commented on gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java:
```diff
@@ -142,7 +150,7 @@ public void init()
     {
         String msgPrefix = "[GEM FM:: " + ((GEMFunctionManager)getUserFunctionManager()).m_FMname
-            + "] GEMEventHandler::GEMEventHandler(): ";
+	+ "] GEMEventHandler::GEMEventHandler(): ";
```
Do you have some indentation handling in your text editor?
We should sync so that we don't have commits that are full of whitespace changes
Once we're settled, I can use my config to open the files and automatically convert things to our desired format, and then we do a `git diff -wb` to see that it only changed the whitespace.

jsturdy commented:
Also regarding:
> If it is local, a `SID` is created, if not `SID=0`
does the FM get the `SID` from the `levelZero` in this case?

jotadram6 commented:
I think it is the case. For global and minidaq is the Lv0 that creates the
SID.

Also, I didn't understand your comment regarding:

>  import java.util.Random;
+import java.net.URL;


Could you tell me what you want me to remove?


However, before accepting the pull request I need to comment one line, so
please wait until the next commit....

On Thu, Sep 14, 2017 at 4:04 PM, Jared Sturdy <notifications@github.com>
wrote:

> Also regarding:
>
> If it is local, a SID is created, if not SID=0
> does the FM get the SID from the levelZero in this case?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-gem-daq-project/GEM-CMS-FM/pull/12#issuecomment-329492260>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AApYfDr5g61Cwq49gTNtOm17mcS33YyMks5siTJvgaJpZM4PUqV7>
> .
>



-- 
_______________________________
José David Ruiz Álvarez
CMS Group - Universidad de los Andes
CERN - Office: 42 - R - 036
_______________________________

jsturdy commented:
Yeah, github isn't great about showing which lines the comments are associated with...
But I also think I was tired and didn't see the difference between the `URI` and `URL` import ;)

jsturdy commented on gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java:
```diff
@@ -2441,8 +2438,8 @@ public void run() {
-				if (status.equals("Failed") || status.equals("Faulty") || status.equals("Error") || stateName.equalsIgnoreCase("failed")) {
-				    String errMessage = "[GEM " + m_gemFM.m_FMname + "] Error! GEMSupervisorWatchThread(): supervisor reports partitionState: " + status + " and stateName: " + stateName +"; ";
+				if (status.equals("Failed") || status.equals("Faulty") || status.equals("Error")) {
+				    String errMessage = "[GEM " + m_gemFM.m_FMname + "] Error! GEMSupervisorWatchThread(): supervisor reports partitionState: " + status +"; ";
```
All the `[GEM " + m_gemFM.m_FMname + "] ` should be replaced with the appropriate `msgPrefix` for uniformity

jsturdy replied:
(of course, where possible, as I had done in constructors and starts of functions)

jsturdy commented on gemLevelOneFM/src/rcms/fm/app/gemfm/GEMFunctionManager.java:
```diff
@@ -481,11 +484,16 @@ protected void sendCMSError(String errMessage)
+	try {
+	    theErrorHandler.setError(error);
+        } catch (Exception e) {
+            logger.warn(msgPrefix + "" + getClass().toString() + ": Failed to send error message " + errMessage);
```
What's the point of the ` + "" + ` here?

jotadram6 commented:
I have addressed the comments,could you please check? I also implemented a
masking of tcds applications in the FM to see if it solves the current
double-controlling issue. Please wait that I test this masking before
accepting the PR.

On Tue, Sep 19, 2017 at 3:29 PM, Jared Sturdy <notifications@github.com>
wrote:

> *@jsturdy* requested changes on this pull request.
> ------------------------------
>
> In gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java
> <https://github.com/cms-gem-daq-project/GEM-CMS-FM/pull/12#discussion_r139691386>
> :
>
> > @@ -2441,8 +2438,8 @@ public void run() {
>  				    m_gemFM.goToError(errMessage);
>  				}
>
> -				if (status.equals("Failed") || status.equals("Faulty") || status.equals("Error") || stateName.equalsIgnoreCase("failed")) {
> -				    String errMessage = "[GEM " + m_gemFM.m_FMname + "] Error! GEMSupervisorWatchThread(): supervisor reports partitionState: " + status + " and stateName: " + stateName +"; ";
> +				if (status.equals("Failed") || status.equals("Faulty") || status.equals("Error")) {
> +				    String errMessage = "[GEM " + m_gemFM.m_FMname + "] Error! GEMSupervisorWatchThread(): supervisor reports partitionState: " + status +"; ";
>
> All the [GEM " + m_gemFM.m_FMname + "] should be replaced with the
> appropriate msgPrefix for uniformity
> ------------------------------
>
> In gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java
> <https://github.com/cms-gem-daq-project/GEM-CMS-FM/pull/12#discussion_r139691522>
> :
>
> > @@ -2441,8 +2438,8 @@ public void run() {
>  				    m_gemFM.goToError(errMessage);
>  				}
>
> -				if (status.equals("Failed") || status.equals("Faulty") || status.equals("Error") || stateName.equalsIgnoreCase("failed")) {
> -				    String errMessage = "[GEM " + m_gemFM.m_FMname + "] Error! GEMSupervisorWatchThread(): supervisor reports partitionState: " + status + " and stateName: " + stateName +"; ";
> +				if (status.equals("Failed") || status.equals("Faulty") || status.equals("Error")) {
> +				    String errMessage = "[GEM " + m_gemFM.m_FMname + "] Error! GEMSupervisorWatchThread(): supervisor reports partitionState: " + status +"; ";
>
> (of course, where possible, as I had done in constructors and starts of
> functions)
> ------------------------------
>
> In gemLevelOneFM/src/rcms/fm/app/gemfm/GEMFunctionManager.java
> <https://github.com/cms-gem-daq-project/GEM-CMS-FM/pull/12#discussion_r139691773>
> :
>
> >              getParentErrorNotifier().sendError(error);
>          } catch (Exception e) {
>              logger.warn(msgPrefix + "" + getClass().toString() + ": Failed to send error message " + errMessage);
> -        }
> +	    }*/
> +	try {
> +	    theErrorHandler.setError(error);
> +        } catch (Exception e) {
> +            logger.warn(msgPrefix + "" + getClass().toString() + ": Failed to send error message " + errMessage);
>
> What's the point of the + "" + here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-gem-daq-project/GEM-CMS-FM/pull/12#pullrequestreview-63658345>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AApYfI5Vy-3H8KxsIEMJNW5-rXucekbLks5sj8HUgaJpZM4PUqV7>
> .
>



-- 
_______________________________
José David Ruiz Álvarez
CMS Group - Universidad de los Andes
CERN - Office: 42 - R - 036
_______________________________

jsturdy commented on gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java:
```diff
@@ -113,11 +113,13 @@
      */
     public boolean stopGEMSupervisorWatchThread =  false; 
 
+    public String msgPrefix = "[GEM FM] GEMEventHandler::GEMEventHandler(): ";
```
This isn't quite what I was talking about...
I wanted the `msgPrefix` to tell you which FM you're running from, and which function, so where it's everywhere else defined at the beginning of the function definition.
I can go through this later and fix it though

jsturdy commented on gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java:
```diff
@@ -2006,6 +2094,32 @@ protected void initXDAQ()
+                msg = msgPrefix + "Masking the  application with name "
+                    + qr.getName() + " running on host " + qr.getResource().getHostName();
+                logger.info(msg);
+                qr.setActive(false);
```
I don't think this is sufficient, as the `Halting` is happening [here](https://github.com/jotadram6/GEM-CMS-FM/blob/db8f7653a616ace97870311dbb8d406b015b0094/gemLevelOneFM/src/rcms/fm/app/gemfm/GEMEventHandler.java#L465)
What should happen is that one should check if the FM is controlling the `tcds` applications, and if not, then it doesn't fill the `tcds` containers earlier
